Feature/camera/#86bxcjezz

OriolGallego2001 opened the pull request:
No warnings or memory leaks.
Camera component movement, destruction, and preview.

M1R4B3L committed:
solved Focus and Orbit

M1R4B3L committed:
Merge branch 'develop' into feature/camera/#86bxcjezz

M1R4B3L committed:
Fixed Orbit when rotating and removed camera utils

M1R4B3L committed:
Camera retouches

M1R4B3L committed:
Merge branch 'develop' into feature/camera/#86bxcjezz

M1R4B3L committed:
Camera with a component

M1R4B3L committed:
Camera

M1R4B3L committed:
Merge branch 'develop' into feature/camera/#86bxcjezz

OriolGallego2001 committed:
Camera with transform

OriolGallego2001 committed:
Fix for avoiding double update movement and transform

OriolGallego2001 committed:
Frustum culling in camera component

OriolGallego2001 committed:
Merge branch 'develop' into feature/camera/#86bxcjezz

OriolGallego2001 committed:
Fixes

OriolGallego2001 committed:
Warning and memory leaks fixed

upcgames2324 commented:
Task linked: [CU-86bxcjezz Camera](https://app.clickup.com/t/86bxcjezz)

carlosgspsp commented on Engine/Source/InspectorPanel.cpp:
```diff
@@ -167,9 +168,33 @@ void InspectorPanel::DrawTransform(GameObject* object) {
 			}
 
 			if (modifiedTransform) {
+				float3 rotation_diff = DegToRad(newRotation) - object->GetRotation();
+				LOG("Rotation");
+				LOG("%f, %f, %f", rotation_diff.x, rotation_diff.y, rotation_diff.z);
+
 				object->SetPosition(newPosition);
 				object->SetRotation(DegToRad(newRotation));
 				object->SetScale(newScale);
+				CameraComponent* cameraComp = static_cast<CameraComponent*>(object->GetComponent(ComponentType::CAMERA));
+				if (cameraComp)
+				{
+					float pos_angle_rotated = rotation_diff.MaxElement();
+					float neg_angle_rotated = rotation_diff.MinElement();
+					float angle_rotated = 1.0;
+					bool has_rotated = false;
+					if (pos_angle_rotated != 0) {
+						angle_rotated = pos_angle_rotated;
+						has_rotated = true;
+					}
+					else if (neg_angle_rotated != 0) {
+						angle_rotated = neg_angle_rotated;
+						has_rotated = true;
+					}
+
+					cameraComp->SetPos(newPosition);
+					if(has_rotated)
+						cameraComp->Rotate(rotation_diff/angle_rotated, angle_rotated);
+				}
```
I think this code should be in the CameraComponent Update method and should be inside an if(IsTransformModified). 

PD: IsTransformModified() method do not exist right now.

OriolGallego2001 replied:
I will have a better look at it, I think this code relates to when the inspector modifies the transform values so it should be in the inspector panel cpp if I find something there unrelated to the inspector I will move it to the proper class.

carlosgspsp commented on Engine/Source/CameraComponent.cpp:
```diff
@@ -34,23 +68,83 @@ Component* CameraComponent::Clone(GameObject* owner) const
 
 void CameraComponent::Reset()
 {
-	mCamera = CameraUtils::InitiateCamera(float3(0.0f, 0.0f, 0.0f));
+	//mCamera = CameraUtils::InitiateCamera(float3(0.0f, 0.0f, 0.0f));
```
This method should be implemented.

OriolGallego2001 replied:
On it.

carlosgspsp commented on Engine/Source/CameraComponent.cpp:
```diff
@@ -34,23 +68,83 @@ Component* CameraComponent::Clone(GameObject* owner) const
 }
 
 void CameraComponent::SetPosition(const float3& position)
 {
-	CameraUtils::CameraComponentTransform(position, *mCamera);
+	//CameraUtils::CameraComponentTransform(position, *mCamera);
 }
```
Because the camera position is attached directly to the gameobject position I think this method should not exist.

OriolGallego2001 replied:
Agree unless you want to force it in scripting, I will implement the method just in case.

carlosgspsp replied:
In scripting to move the camera you move the gameobject attached to the camera like in Unity, this method should not exist.

carlosgspsp commented on Engine/Source/CameraComponent.cpp:
```diff
@@ -34,23 +68,83 @@ Component* CameraComponent::Clone(GameObject* owner) const
 void CameraComponent::SetRotation(const float3& rotation)
 {
-	CameraUtils::Rotate(rotation,-1 ,*mCamera);
+	//CameraUtils::Rotate(rotation,-1 ,*mCamera);
```
This comment should be deleted

OriolGallego2001 replied:
👍

carlosgspsp commented on Engine/Source/ModuleCamera.h:
```diff
@@ -7,30 +7,40 @@
 
-class Quadtree;
-class GameObject;
+#include "CameraComponent.h"
+#include "GameObject.h"
```
I think these could be a forward declaration

OriolGallego2001 replied:
Then the getters of the camera component will be moved to the .cpp

carlosgspsp replied:
Yeah, i think it would be better, also those methods should not exist in the next versions.

carlosgspsp commented on Engine/Source/InspectorPanel.cpp:
```diff
@@ -610,46 +635,48 @@ void InspectorPanel::DrawCameraComponent(CameraComponent* component)
-	float* FarBar = &Far;
-	const char* FarLabel = "FarPlane";
+	static float farPlane = component->GetFarPlane();
+	const char* farLabel = "FarPlane";
```
Add ## to all the labels here so they are not shown in the inspector two times like "##FarPlane"

OriolGallego2001 replied:
I don't understand this comment😕 when I execute the engine the labels are not shown twice...

carlosgspsp replied:
![image](https://github.com/RollingBarrel/Helheim-Engine/assets/46396509/65a8a17b-814c-45c8-aba6-66de26e95df3)

carlosgspsp commented on Engine/Source/CameraComponent.cpp:
```diff
@@ -34,23 +68,83 @@ Component* CameraComponent::Clone(GameObject* owner) const
 
 void CameraComponent::Save(Archive& archive) const
 {
-	//archive.AddInt("ID", mID);
-	//archive.AddInt("Camera", mCamera);
+    //archive.AddInt("ID", mID);
+    //archive.AddInt("Camera", mCamera);
 }
```
This method should be implemented in order to save camera values (near plane, far plane and FOV) in the scene.

OriolGallego2001 replied:
Will add a task to click up for the next task, but maybe this PR can pass without it?

carlosgspsp replied:
Okay try to have this for the next PR son gameplay team can set camera values between sessions.👍

carlosgspsp commented on Engine/Source/CameraComponent.h:
```diff
@@ -18,19 +18,36 @@ class ENGINE_API CameraComponent :
+    Frustum mFrustum;
+    bool mEnableCulling = true;
+
+    float mAspectRatio;
```
Initialize this variable

OriolGallego2001 replied:
It is done in the .cpp as it needs values from the frustum. No warnings should arise because of this.

carlosgspsp commented on Engine/Source/InspectorPanel.cpp:
```diff
@@ -610,46 +635,48 @@ void InspectorPanel::DrawCameraComponent(CameraComponent* component)
+	if(ImGui::Button("Make Current Camera"))
 	{
-		component->SetNearPlane(Near);
-		component->SetFarPlane(Far);
-		component->SetVerticicalFOV(DegToRad(FOV));
+		App->GetCamera()->SetCurrentCamera(const_cast<GameObject*>(component->GetOwner()));
 	}
```
When I click this button how can I change back to the editor camera?

OriolGallego2001 replied:
I can implement it but as make current is a debug feature when using the engine it should not matter much.

carlosgspsp replied:
There should be an option to return to the editor camera so gameplay tema can debug everything more easily

OriolGallego2001 replied:
Done

carlosgspsp commented on Engine/Source/ModuleCamera.h:
```diff
@@ -7,30 +7,40 @@
+
+	void SetCurrentCamera(GameObject* camera);
+
+	const CameraComponent* GetCurrentCamera() { return (CameraComponent*)mCurrentCamera->GetComponent(ComponentType::CAMERA); }
+
+	const CameraComponent* GetEditorCamera() { return (CameraComponent*)mEditorCamera->GetComponent(ComponentType::CAMERA); }
```
The const next to () is missing. "const CameraComponent* GetCurrentCamera() **const** {...}"

OriolGallego2001 replied:
true

OriolGallego2001 committed:
Carlos comments were taken into consideration part1

Parent update not in inspector but inside component update.
Inspector fixes.
Other methods.

OriolGallego2001 committed:
Test for scripting

OriolGallego2001 committed:
Fixing the editor camera v1

carlosgspsp committed:
Rotation

carlosgspsp committed:
Polish

OriolGallego2001 committed:
Better code

OriolGallego2001 committed:
Merge branch 'feature/camera/#86bxcjezz' of https://github.com/RollingBarrel/Helheim-Engine into feature/camera/#86bxcjezz
